scripts/generate_rust_analyzer.py

```diff
@@ -98,13 +99,15 @@ def append_crate(display_name, root_module, deps, cfg=[], is_workspace_member=Tr
     # Then, the rest outside of `rust/`.
     #
     # We explicitly mention the top-level folders we want to cover.
-    for folder in ("samples", "drivers"):
+    extra_src_dirs = ["samples", "drivers"] if external_src is None else [external_src]
+
+    for folder in extra_src_dirs:
         for path in (srctree / folder).rglob("*.rs"):
             logging.info("Checking %s", path)
             name = path.name.replace(".rs", "")
 
             # Skip those that are not crate roots.
-            if f"{name}.o" not in open(path.parent / "Makefile").read():
+            if os.path.exists(path.parent / "Makefile") and f"{name}.o" not in open(path.parent / "Makefile").read():
```

Comment:
As I was testing this with the rust-out-of-tree-module example, I found that the crate was getting skipped because of this logic.  I believe this will also need to search a possible `Kbuild` file (E.g. [the example](https://github.com/Rust-for-Linux/rust-out-of-tree-module/blob/main/Kbuild))

```
diff --git a/scripts/generate_rust_analyzer.py b/scripts/generate_rust_analyzer.py
index 1546b80db554..f5567825c94c 100755
--- a/scripts/generate_rust_analyzer.py
+++ b/scripts/generate_rust_analyzer.py
@@ -107,7 +107,15 @@ def generate_crates(srctree, objtree, sysroot_src, external_src):
             name = path.name.replace(".rs", "")
 
             # Skip those that are not crate roots.
-            if os.path.exists(path.parent / "Makefile") and f"{name}.o" not in open(path.parent / "Makefile").read():
+            def is_crate_root(*filenames):+                for filename in filenames:
+                    filepath = path.parent / filename
+                    if os.path.exists(filepath) and f"{name}.o" in open(filepath).read():
+                        return True
+                return False
+
+            if not is_crate_root("Makefile", "Kbuild"):
+                logging.warning(f"skipping")
                 continue
 
             logging.info("Adding %s", name)
```